design/external-references.md

```diff
@@ -0,0 +1,394 @@
+
+```ts
+const vpc = VpcNetwork.lookupVpc(this, {
+  tags: {
+    department: 'sales',
+    stage: 'prod'
+  }
+});
+```
+
+The underlying implementation here is different, it uses an environmental context provider to lookup the VPC and extract all the relevant information from it, such as subnets, NAT Gateways and route tables.
+
+Here too, we expect idempotent behavior, which can be implemented in a similar manner.
+
+## Other Applications
+
+The construct serialization mechanism opens an opportunity for other applications that may benefit from being able to reference CDK constructs from outside the app. This section describes a few examples.
+
+### Serialization to JSON
+
+It should be trivial to implement a JSON serialization context:
+
+```ts
+const ctx = new JsonSerializtionContext();
+alb.serialize(ctx);
+
+assertEquals(resolve(ctx.json), {
+  "loadBalancerArn": { "Fn::GetAtt": [ "MyALB1288xxx", "Arn" ] },
+  "securityGroup": {
+    "securityGroupId": { "Ref": "MyALBSecurityGroup4444" }
+  }
+});
+```
+
+Representing a construct's runtime attributes as JSON (or a stringified JSON if needed via `CloudFormationJSON`) opens up a few interesting applications as described below.
+
+### Cross-Account/Region References
+
+The serialization mechanism, together with a bunch of custom resources can be used to reference constructs across accounts and region via, e.g. an S3 bucket.
+
+The producing stack can write a file to an S3 bucket with e.g. the JSON serialized representation of the construct and the consuming stack can read this file during deployment and deserialize the construct.
+
+### Environment Variables Serialization
+
+When runtime code needs to interact with resources defined in a CDK app, it needs to be able to reference these resources.
+
+The current practice is to manually wire specific resource attributes via environment variables so they will be available for runtime code. This may be sufficient for simple use cases such as simple AWS resources where a single attribute might be sufficient to represent the construct, but more complex scenarios (such as composite constructs) may benefit from the ability to serialize the entire construct through environment variables either through individual keys or as a single key + JSON value.
+
+## Implementation Notes
+
+The underlying pattern we use today for supporting imports/exports (`IBucket`, `BucketBase`, `Bucket` and `ImportedBucket`) continues to be relevant for implementing serialization and the `fromXxx` methods.
+
+The various static import methods (`deserializeXxx` `importXxx`, `fromXxx`) will all return an object that implements `IXxx`. The concrete type of this object will be an internal class `ImportedXxx` that includes the heuristics of how to represent an external resource.
```

Comment:
I don't think `heuristics` are applied here. Also, I don't think the "concrete type" of the returned object needs to be specified here.

Reply:
It's under "implementation notes". I will make it clear that this is not a requirement but rather a recommendation.